.github/CONTRIBUTING.md

```diff
@@ -53,68 +57,86 @@ issue. Stale issues will be closed.
 
 [./docs]: ../docs/README.md
 [pkg.go.dev/github.com/hashicorp/consul]: https://pkg.go.dev/github.com/hashicorp/consul
 
-### Checklists
+### Testing
+
+During development, it may be more convenient to check your work-in-progress by running only the tests which you expect to be affected by your changes, as the full test suite can take several minutes to execute. [Go's built-in test tool](https://golang.org/pkg/cmd/go/internal/test/) allows specifying a list of packages to test and the `-run` option to only include test names matching a regular expression.
+The `go test -short` flag can also be used to skip slower tests.
+
+Examples (run from the repository root):
+- `go test -v ./connect` will run all tests in the connect package (see `./connect` folder)
+- `go test -v -run TestRetryJoin ./command/agent` will run all tests in the agent package (see `./command/agent` folder) with name substring `TestRetryJoin`
 
-Some common changes that many PRs require such as adding config fields, are
-documented through checklists.
+When a pull request is opened CI will run all tests and lint to verify the change.
 
-Please check in [docs/](../docs/) for any `checklist-*.md` files that might help
-with your change.
+### Submitting a Pull Request
+
+Before writing any code, we recommend:
+- Create a Github issue if none already exists for the code change you'd like to make.
+- Write a comment on the Github issue indicating you're interested in contributing so
+maintainers can provide their perspective if needed.
+
+When you're ready to submit a pull request:
+1. Review the [list of checklists](#checklists) for common changes and follow any
+   that apply to your work.
+2. Include evidence that your changes work as intended (e.g., add/modify unit tests).
+3. Open the PR from your fork against base repository `hashicorp/consul` and branch `main`.
+   - [Link the PR to its associated issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
+4. Include any specific questions that you have for the reviewer in the PR description
+   or as a PR comment in Github.
+   - If there's anything you find the need to explain or clarify in the PR, consider
+   whether that explanation should be added in the source code as comments.
+   - You can submit a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)
+   if your changes aren't finalized but would benefit from in-process feedback.
+5. If there's any reason Consul users might need to know about this change,
+   [add a changelog entry](../docs/contributing/changelog-entry.md).
+6. After you submit, the Consul maintainers team needs time to carefully review your
+   contribution and ensure it is production-ready, considering factors such as: security,
+   backwards-compatibility, potential regressions, etc.
+7. After you address Consul maintainer feedback and the PR is approved, do the honors
+   of clicking the Github merge button!
```

Comment:
Outside constributors can not click the button. Only someone with write permission on the repo can. So maybe this should say "the PR will be merged once it is approved" ?

Reply:
Good point - I'll fix.

Should we ask the contributor to indicate if, in their opinion, it is ready to merge before we click it ourselves? In my experience, one of the points of letting the PR author click merge is so they can confirm everything is actually done (whereas the reviewer might do so prematurely). Is this a real concern? (In some cases, we might know it's done, but in others we may not be sure.)

Reply:
I think for community PRs we are cautious enough that we can make that decision. If they want to add something it can always be a follow up PR. I'm weary of waiting around for them to respond as another step in the process.

Reply:
@dnephin : changing to the below. Let me know what you think.
```
7. After you address Consul maintainer feedback and the PR is approved, a Consul maintainer
   will merge it. Your contribution will be available from the next major release (e.g., 1.x)
   unless explicitly backported to an existing or previous major release by the maintainer.
   ```

Reply:
Sounds great!